### DISCUSSION TITLE
Authentication - Retrieve user tokens independently from its provider

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#17963] [Archive]
Hi,

Before implementing this feature ( https://github.com/directus/directus/pull/13184) ensuring the user provider match, we were able to get tokens for Any user independently from its provider. We were able to retrieve tokens using the `DEFAULT_AUTH_PROVIDER` and the user email for any users. It was really usefull to be able to reuse login function logic with a custom auth logic.

It seems the purpose of this feature was more to give more feedback to the user than to make the login function more rigid.

Here is the check: https://github.com/directus/directus/blob/75ad4876bfecc2f8cdac910233396f6a79fef7b4/api/src/services/authentication.ts#L130-L133

Because the login function aims to retrieve the tokens for a given email could we make it less rigid and still allow one to retrieve tokens for Any provider just by using the email.
https://github.com/directus/directus/blob/75ad4876bfecc2f8cdac910233396f6a79fef7b4/api/src/services/authentication.ts#L40-L50

Here is some suggestions:

#1 We could alter the check to performs it on the updated payload instead of checking the user provider. It would make sense to be able to alter the provider in the login function.
```
const updatedPayload = await emitter.emitFilter('auth.login', payload,...)
```
then
```
else if (updatedPayload.provider !== providerName) {
	await stall(STALL_TIME, timeStart);
	throw new InvalidProviderException();
}
```

#2 Make this test optionnal

I'm take the liberty of quoting you @aidenfoxx as you're the authentication specialist.

Thanks for feedback

**Comment by aidenfoxx** [comment]
In regards to that PR, the check was not added, but moved. Previously it was the responsibility of the SQL query to determine if the user belonged to the correct provider: https://github.com/directus/directus/pull/13184/files#diff-03b58901729f938551b47a9357d79ae1d0c62868265a05e8f9980d30b22c0e95L71

The check exists for an important reason. In a controlled auth environment you would not want your users being able to skirt around your authentication provider by setting themselves a password and logging in via the email/password flow.

There are also other things to consider, such as the fact that we store the OAuth "refresh_token" with the user data, and if a user could log in with multiple OAuth providers, I imagine it could cause issues.

These things are solvable, of course. But I don't think there is one simple thing we can do to just enable the functionality in a consistently good way.

> I'm take the liberty of quoting you @aidenfoxx as you're the authentication specialist.

I appreciate the confidence, but I'm a authentication enthusiast at best!

**Reply by gkielwasser** [reply]
I see your point. I think the issue is a lot of logic is encapsulted in the login function and we lack of some utils functions that could be reused to allow custom extension not reinvent the wheel concerning some steps around login in Directus. 

For instance, the token generation part could be extracted and exposed in an util function.

After 2 years trying not to duplicate the login function I finally have duplicated 75% of the login function to be able to generate tokens based on a user which have been authentified in native application from native sdks. The percentage could have been lower but the idea was to stick with the differents steps done in the login function,

Not sure how to decouple this method but for instance those steps could be extracted:
- forge of refreshToken & refreshTokenExpiration
- insert a new directus_session, delete old ones
- create activity based on accountability
- update of the last user access

Basically a first step would be to extract all the steps done after user is considered logged that are leading to the generation of the tokens.

@rijkvanzanten Even if you have changed your mind in the meantime, you previously mention something similar here https://github.com/directus/directus/issues/12574#issuecomment-1090726185

**Reply by rijkvanzanten** [reply]
I think the ultimate goal is to have these auth providers be individual packages (that can be loaded in as extensions), and (to your point) expose a set of often used functions as utilities
